### PR TITLE
Add BitBucket Server Project

### PR DESCRIPTION
**Pull request by fejal**
## Contribution type

New project type to support Bit Bucket Server in php-censor.

## Description of change

This adds a preliminary project type that handles the following:

1. Allows setting Project to Bitbucket Server.
2. Assume basic Git can be used to access clones from Bit Bucket server.
3. Sets up the listener in web hooks to fire a build pull requests when they are created in Bit Bucket server.

Known limitations:
1. Does not get list of all commits included in a pull request.
2. Clickable links in the header of the build screen do not map back to the bit bucket server when clicked. Need another configuration field in project setup to get this as the URL to the server will likely be different than the url to the repository on a typical bit bucket server setup.

### CODE REVIEW

**Comment by corpsee** [comment]
@fejal Good work! I will review the PR soon when I have free time. Thanks! :+1:

File: src/Model/Base/Project.php [hunk 1]
```diff
@@ -15,6 +15,7 @@ class Project extends Model
     const TYPE_GOGS         = 'gogs';
     const TYPE_HG           = 'hg';
     const TYPE_BITBUCKET_HG = 'bitbucket-hg';
+    const TYPE_BITBUCKET_SV = 'bitbucket-server';
```
Comment (corpsee) [inline src/Model/Base/Project.php]: @fejal Could you rename const `TYPE_BITBUCKET_SV` to `TYPE_BITBUCKET_SERVER` please?

File: src/Helper/CommandExecutor.php [hunk 2]
```diff
@@ -147,7 +147,7 @@ function ($a) {
 
         $shouldOutput = ($this->logExecOutput && ($this->verbose || 0 !== $status));
 
-        if ($shouldOutput && !empty($this->lastOutput)) {
+        if (($shouldOutput || true) && !empty($this->lastOutput)) {
```
Comment (corpsee) [inline src/Helper/CommandExecutor.php]: Why?

File: src/Logging/BuildDBLogHandler.php [hunk 3]
```diff
@@ -71,7 +71,7 @@ protected function write(array $record)
     {
         $message = (string)$record['message'];
         $message = str_replace(['\/', '//'], '/', $message);
-        $message = str_replace($this->build->getBuildPath(), '<BUILD_PATH>/', $message);
+        //$message = str_replace($this->build->getBuildPath(), '<BUILD_PATH>/', $message);
```
Comment (corpsee) [inline src/Logging/BuildDBLogHandler.php]: Why?

File: src/Model/Build/BitbucketServerBuild.php [hunk 4]
```diff
@@ -0,0 +1,225 @@
+<?php
+
+namespace PHPCensor\Model\Build;
+
+use GuzzleHttp\Client;
```
Comment (corpsee) [inline src/Model/Build/BitbucketServerBuild.php]: @fejal Could you remove useless namespaces (`GuzzleHttp\Client`, `PHPCensor\Config`, `PHPCensor\Model\BuildError`) please?

File: src/Model/Build/BitbucketBuild.php [hunk 5]
```diff
@@ -295,40 +295,4 @@ public function reportError(
 
         parent::reportError($builder, $plugin, $message, $severity, $file, $lineStart, $lineEnd);
     }
-
```
Comment (corpsee) [inline src/Model/Build/BitbucketBuild.php]: Why? Method `getDiffLineNumber` used in method `reportError` (line `273`).

**Comment by corpsee** [comment]
#260.

File: src/Model/Base/Project.php (hunk 1 above)
Comment (fejal) [reply src/Model/Base/Project.php]: Yes, will do.

File: src/Helper/CommandExecutor.php (hunk 2 above)
Comment (fejal) [reply src/Helper/CommandExecutor.php]: Oops, debugging.   Need to remove that.

File: src/Logging/BuildDBLogHandler.php (hunk 3 above)
Comment (fejal) [reply src/Logging/BuildDBLogHandler.php]: Had trouble setting the build path correctly but found the trouble was in how I set it in the config, not in the code.   I'll remove the comment line.

File: src/Model/Build/BitbucketServerBuild.php (hunk 4 above)
Comment (fejal) [reply src/Model/Build/BitbucketServerBuild.php]: Absolutely.

File: src/Model/Build/BitbucketBuild.php (hunk 5 above)
Comment (fejal) [reply src/Model/Build/BitbucketBuild.php]: Oops removed that function from the wrong file when I was cleaning up the code for the initial push, will fix that.

Comment (fejal) [reply src/Model/Build/BitbucketBuild.php]: When cleaning up code for initial commit, (for functionality not implemented in this first patch) I removed getDiffLineNumber from the wrong file it appears.

**Comment by corpsee** [comment]
@fejal Thank you! Good work. I merged your PR manually to `master` branch. The PR will be released with next minor release `1.1.0`.

**Pull request closed without merging**